### PR TITLE
[CK_TILE] Support WMMA (gfx12) in FMHA

### PR DESCRIPTION
**Pull request by ex-rzr**
## Proposed changes

Changes worth mentioning:
* The split KV combine kernel is more universal now, it supports any power-of-2 number of warps, any number of splits and kN1 = 16, see https://github.com/ROCm/composable_kernel/pull/2528/commits/e08a84d56d7babdccb4d42e9b3b7262c45678b11
* Some of hardcoded values (M0 of the dot bwd kernel, N1 of the split KV combine kernel) are set based on tile size, N1 of the split KV combine kernel is ensured to be consistent with the split KV kernel.
* Instances (tile size + pipeline) are checked for duplications and proper ordering of paddings (from non-padded to fully padded).
* A workaround for a compiler bug (SWDEV-559729) in bwd kernels with padding.
* WarpGemmDispatcher is used when possible instead of a concrete warp gemm implementation.
* Some tile distributions are updated to support different tile sizes and warp size.
* Initially this PR included a big change for supporting 16x16 warp tiles in `BlockDropout`, they were merged to develop as a part of #2836.

The main change is about supporting different architectures in one binary (for example, tests and examples) without a need to compile separate binaries for WMMA+warpsize32 and MFMA+warpsize64.

I extended the `KernelComponentFactory` mechanism, already existing in fwd, now all APIs work in this way:
* There are different factories for gfx12 and gfx9, with different tile sizes and pipeline parameters.
* There is also a factory for gfx950 based on gfx9 which adds pipelines with transpose loading (where suitable).
* Later additional factories can be added (for example, for gfx1250).
* Cmake scripts pass the list of targets to python scripts that generate FMHA kernels.
* Each kernel file is supposed to be built for one architecture "family" (gfx12, gfx9, gfx950), file names include corresponding suffixes.
* Building for other architectures is disabled by a special preprocessor guard.
* The `fmha_*_api.cpp` files check what the current device is and launches the required kernel, if there is no kernel for the current device (only for other architectures), -1 is returned, aka "not supported yet" instead of a crash with `hipErrorNoBinaryForGpu`.
* Each kernel includes a special tag for architecture, I added `gfx9_t` and `gfx950_t` to existing `gfx11_t` and `gfx12_t`. This is needed to link two kernels with the same template parameters built for different architectures (otherwise the linker keeps only one of the symbols with the same symbol name).
* Each host dispatching function (like `fmha_fwd_`) has a new template parameter with the architecture tag for a similar reason: having two host functions with the same traits built in files for different architectures leads to duplication symbols. These tags are optional and, in my opinion, are backward compatible.

`fmha_fwd_d..._gfx12.cpp`:
```cpp
...
#if !defined(__HIP_DEVICE_COMPILE__) || defined(__gfx12__)

using fmha_arch_tag = ck_tile::gfx12_t;
...
template<>
float fmha_fwd_<trait_0, fmha_arch_tag>(const ck_tile::stream_config& s, fmha_fwd_args a)
{
...
}
#endif // !defined(__HIP_DEVICE_COMPILE__) || defined(__gfx12__)
```

`fmha_fwd_api.cpp`:
```cpp
float fmha_fwd(fmha_fwd_traits t, fmha_fwd_args a, const ck_tile::stream_config& s) {
...
    [[maybe_unused]] const std::string device_name = ck_tile::get_device_name();

    if(device_name.compare(0, 6, "gfx950") == 0) {
        if(t.data_type.compare("fp16") == 0) {
            if(t.hdim_q <= 32 && t.hdim_v <= 32) {
                ...
                    return fmha_fwd_<trait_, ck_tile::gfx950_t>(s, a);
                ...
    }
    else if(device_name.compare(0, 5, "gfx12") == 0) {
        ...
    else if(device_name.compare(0, 4, "gfx9") == 0) {
        ...
    }
}
```

Here are examples of the generated files:
* [fmha_fwd_d256_bf16_group_b64x64x32x256x32x256_r4x1x1_r4x1x1_w16x16x16_w16x16x16_qr_vr_psskddv_nlogits_alibi_nmask_nlse_ndropout_nskip_nsquant_ntrload_gfx12.cpp](https://github.com/user-attachments/files/22883415/fmha_fwd_d256_bf16_group_b64x64x32x256x32x256_r4x1x1_r4x1x1_w16x16x16_w16x16x16_qr_vr_psskddv_nlogits_alibi_nmask_nlse_ndropout_nskip_nsquant_ntrload_gfx12.cpp)
* [fmha_fwd_api.cpp](https://github.com/user-attachments/files/22883426/fmha_fwd_api.cpp)

Please, let me know if something in this description is not clear.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [x] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [x] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [x] I have removed the stale documentation which is no longer relevant after this pull request
- [x] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

### CODE REVIEW

**Comment by ex-rzr** [comment]
I have added a description of the changes. Ready for review

File: include/ck_tile/core/tensor/buffer_view.hpp
Comment (illsilin) [inline include/ck_tile/core/tensor/buffer_view.hpp:1]: I don't quite understand this part.
The purpose of amd_buffer_addressing_builtins.hpp was to add all the new built-ins supported by newer compiler versions, which will never be available in the older compilers. Hence the #if/#else switch based on the compiler version. Otherwise they should offer identical functionality. So what's the reason to include both of them?

File: include/ck_tile/host/kernel_launch.hpp [hunk 1]
```diff
@@ -28,18 +28,47 @@ __launch_bounds__(Kernel::kBlockSize, MinBlockPerCu)
...
+
 //
 // return a anonymous functor(lambda) to be called later
 // the KernelImpl should be a class without non-static data member, or let's say
 // can be instantiate with "KernelImpl{}"
 //
 // the "static __device__ operator()(some_arg)" is the entry point of KernelImpl
 //
-template <int MinBlockPerCu = CK_TILE_MIN_BLOCK_PER_CU, typename KernelImpl, typename... Args>
+// Arch can be used to support linking multiple object files that have the same kernel compiled for
```
Comment (linqun) [inline include/ck_tile/host/kernel_launch.hpp:51]: what will happen if we don't add arch in kenerl template? link error or something else?
I think compiler should have handled kernel on multi device.

**Review by linqun — COMMENTED** [review]
looks good to me for others.

File: example/ck_tile/01_fmha/script/smoke_test_fwd.sh [hunk 2]
```diff
@@ -114,9 +114,9 @@ run_fp8fp32_tests() {
     for perm in 0 1 ; do
     for bias in "n" "e" "a" ; do
     for b in 1 2 ; do
-    for hdim in 64 128 256 ; do
+    for hdim in 128 ; do
```
Comment (joyeamd) [inline example/ck_tile/01_fmha/script/smoke_test_fwd.sh:117]: him 64, 256 is removed; need to restore.

File: include/ck_tile/core/tensor/buffer_view.hpp
Comment (ex-rzr) [reply include/ck_tile/core/tensor/buffer_view.hpp:1]: The problem is that, despite the name, `amd_buffer_addressing_builtins.hpp` doesn't use builtins, only inline assembly which is not compatible with RDNA4 (different instruction names for loads/stores, 32-bit exec masks etc.).

I don't understand why it's done this way.
And I don't have enough information to remove `amd_buffer_addressing_builtins.hpp`, because it's possible that inline assembly versions are faster in some cases (especially finctions with conditional loads/stores that explicitly set and restore exec mask).

It seems that this is a topic for a separate investigation as it requires performance evaluation of both solutions in real applications. 

In this file I just removed the check of the compiler version to make it consistent with two other places (`core.hpp` and `arch.hpp`) where they are included like this:
```
#include "ck_tile/core/arch/amd_buffer_addressing_builtins.hpp"
#include "ck_tile/core/arch/amd_buffer_addressing.hpp"
```
`amd_buffer_addressing.hpp` and `amd_buffer_addressing_builtins.hpp` have guards with `CK_TILE_USE_BUFFER_ADDRESSING_BUILTIN`, this definintion is set in `include/ck_tile/core/config.hpp`.

So overall I didn't change functionality, just forced to use the version with builtins on >=gfx10 (in `config.hpp`) and removed `__clang_major__ >= 20` as its already controlled by `CK_TILE_USE_BUFFER_ADDRESSING_BUILTIN`.

File: include/ck_tile/host/kernel_launch.hpp (hunk 1 above)
Comment (ex-rzr) [reply include/ck_tile/host/kernel_launch.hpp:51]: Let's assume that we generate the same kernel (same tile size and pipeline) for different architectures, gfx9 and gfx12. Each is built as a separate file: `fmha..._gfx9.o` and `fmha..._gfx12.o`. So these two files use the same symbol for the kernel built for different architectures, the linker does not combine them into one "fat" symbol but keeps only one of them in the resulting binary.

In this case we can get a **runtime** error that there is not kernel for the current device (and on another device it works).

File: test/ck_tile/fmha/test_fmha_fwd.cpp [hunk 3]
```diff
@@ -38,7 +38,7 @@ struct TestConfigs
     static constexpr auto AppendKVHDimValues = std::array{
         std::tuple{32, -1}, std::tuple{64, -1}, std::tuple{128, -1}, std::tuple{256, -1}};
     static constexpr auto ModeValues        = std::array{mode_enum::batch, mode_enum::group};
-    static constexpr auto IsVRowmajorValues = std::array{false, true};
+    static constexpr auto IsVRowmajorValues = std::array{true};
```
Comment (joyeamd) [inline test/ck_tile/fmha/test_fmha_fwd.cpp:41]: Maybe also need to restore back IsVRowmajorValues = std::array{false, true}

File: example/ck_tile/01_fmha/script/smoke_test_fwd.sh (hunk 2 above)
Comment (ex-rzr) [reply example/ck_tile/01_fmha/script/smoke_test_fwd.sh:117]: But there are no instances for hdim != 128: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py#L672

Previously it didn't matter because it had `-d=128` and ignored the actual `hdim`

File: test/ck_tile/fmha/test_fmha_fwd.cpp (hunk 3 above)
Comment (ex-rzr) [reply test/ck_tile/fmha/test_fmha_fwd.cpp:41]: A similar reason, currently there are no instances with `"col"`: https://github.com/ROCm/composable_kernel/blob/develop/example/ck_tile/01_fmha/codegen/ops/fmha_fwd.py#L699

Of course, in this case the test cases with IsVRowmajorValues = false would be marked as SKIPPED, not a big problem

**Comment by ex-rzr** [comment]
It seems that my changes in the codegen scripts are not compatible with aiter and pytorch.
I'm working on this. Perhaps passing `--targets` will be enough, I need to check it.

File: include/ck_tile/core/tensor/buffer_view.hpp
Comment (illsilin) [reply include/ck_tile/core/tensor/buffer_view.hpp:1]: OK, we'll take care of this in a separate PR.

**Review by illsilin — DISMISSED** [review]

**Comment by ex-rzr** [comment]
One of the new commit is mostly refactoring: less error-prone duplication of arch-related formatting templates in every codegen script.

Another moves some code from Jenkins pipeline to a separate function to work around the Jenkins crash with "Method too large" (CI jobs couldn't even started). Apparently merging the recent develop branch with some additions to the pipeline plus my changes in this branch made the pipeline too large to be compiled as a single Java method. My fix works but the issue may return later (There is a JVM property  recommended in https://issues.jenkins.io/browse/JENKINS-37984 as a workaround for a similar issue, I don't know if it helps with this one).

**Review by illsilin — APPROVED** [review]
